In [2]:
import pandas as pd
import seaborn as sns
import numpy as np

df = pd.read_csv('https://raw.githubusercontent.com/fazlyrabbi77/DataProcessing/master/real-estate.csv')
df

PID  ST_NUM     ST_NAME OWN_OCCUPIED NUM_BEDROOMS NUM_BATH SQ_FT
0  100001000.0   104.0      PUTNAM            Y            3        1  1000
1  100002000.0   197.0   LEXINGTON            N            3      1.5    --
2  100003000.0     NaN   LEXINGTON            N          NaN        1   850
3  100004000.0   201.0    BERKELEY           12            1      NaN   700
4          NaN   203.0    BERKELEY            Y            3        2  1600
5  100006000.0   207.0    BERKELEY            Y          NaN        1   800
6  100007000.0     NaN  WASHINGTON          NaN            2   HURLEY   950
7  100008000.0   213.0     TREMONT            Y            1        1   NaN
8  100009000.0   215.0     TREMONT            Y           na        2  1800

We have only eight columns and nine rows in this data set. The purpose of taking this small dataset is to make it easier to understand the use of different functions of data cleaning. This dataset is basically a real estate dataset. Here is some information about the different apartments including the ID number of the apartment, the street number on the road, the street name, the number of bedrooms in the apartment, the number of bathrooms, the number of square feet of the apartment and whether the owner himself lives here. .

### Replacing Columns

In [3]:
#Renaming Columns with inplace
df.rename(columns = {"NUM_BEDROOMS": "BEDROOMS",  "NUM_BATH":"BATH"},inplace=True) 
df.head()

PID  ST_NUM    ST_NAME OWN_OCCUPIED BEDROOMS BATH SQ_FT
0  100001000.0   104.0     PUTNAM            Y        3    1  1000
1  100002000.0   197.0  LEXINGTON            N        3  1.5    --
2  100003000.0     NaN  LEXINGTON            N      NaN    1   850
3  100004000.0   201.0   BERKELEY           12        1  NaN   700
4          NaN   203.0   BERKELEY            Y        3    2  1600

In [7]:
df.dtypes

PID             float64
ST_NUM          float64
ST_NAME          object
OWN_OCCUPIED     object
BEDROOMS         object
BATH             object
SQ_FT            object
dtype: object

It is easy to understand from our dataset that there are several NaN values, but also many Garbage values. In this case this dataset is unsuitable for data modeling. So before we go to the next step, we need to clean and make this dataset useful.
Many times we need to rename the feature or column of the dataset, in which case we can rename the column through the following function. If you do not use inplace = True, it will return to its previous state, so we used inplace = True to save this change.

### Checking NAN values

In [8]:
df.isnull().values.any()

True

In [9]:
df.isnull().values.sum()

8

In [10]:
#Showing the null values as per attributes
null_columns=df.columns[df.isnull().any()]
df[null_columns].isnull().sum()

PID             1
ST_NUM          2
OWN_OCCUPIED    1
BEDROOMS        2
BATH            1
SQ_FT           1
dtype: int64

In [11]:
print(df[df.isnull().any(axis=1)][null_columns].head())

           PID  ST_NUM OWN_OCCUPIED BEDROOMS    BATH SQ_FT
2  100003000.0     NaN            N      NaN       1   850
3  100004000.0   201.0           12        1     NaN   700
4          NaN   203.0            Y        3       2  1600
5  100006000.0   207.0            Y      NaN       1   800
6  100007000.0     NaN          NaN        2  HURLEY   950


### NAN Values Handling

Now it's time to treat the NaN value that we got. We can delete the row of NaN values, but if the dataset is too small we will lose a lot of data, so this is not an effective method. We will try different ways to fill the NaN values ​​with any possible logical value. There are many ways to handle NaN values, we will choose to look at them step by step.
We saw earlier that the PID column has a NaN value. The PID of the previous row has null value of 1000040000 and the PID of the next row is 1000060000. So we understand by common sense that the value of NaN value of this row will be 100005000. We can easily fill in the NaN value of the PID column using the fillna method.
If there were multiple NaN values ​​here, all NaN values ​​would be filled by 100005000 because of this method. We have given this example only to show how NaN value can be filled by a specific value.

In [12]:
df.head(5)

PID  ST_NUM    ST_NAME OWN_OCCUPIED BEDROOMS BATH SQ_FT
0  100001000.0   104.0     PUTNAM            Y        3    1  1000
1  100002000.0   197.0  LEXINGTON            N        3  1.5    --
2  100003000.0     NaN  LEXINGTON            N      NaN    1   850
3  100004000.0   201.0   BERKELEY           12        1  NaN   700
4          NaN   203.0   BERKELEY            Y        3    2  1600

In [16]:
df['PID'].dtype

dtype('float64')

1. Drop total null values
2. Fill out with the median values
3. Fill out with the mode values

In [7]:
# Filling null values with specific value
df['PID'].fillna(100005000,inplace=True)
df

PID  ST_NUM     ST_NAME OWN_OCCUPIED BEDROOMS    BATH SQ_FT
0  100001000.0   104.0      PUTNAM            Y        3       1  1000
1  100002000.0   197.0   LEXINGTON            N        3     1.5    --
2  100003000.0     NaN   LEXINGTON            N      NaN       1   850
3  100004000.0   201.0    BERKELEY           12        1     NaN   700
4  100005000.0   203.0    BERKELEY            Y        3       2  1600
5  100006000.0   207.0    BERKELEY            Y      NaN       1   800
6  100007000.0     NaN  WASHINGTON          NaN        2  HURLEY   950
7  100008000.0   213.0     TREMONT            Y        1       1   NaN
8  100009000.0   215.0     TREMONT            Y       na       2  1800

In our dataset the ST_NUM column has two NaN values. We will now see how to fill the NaN value based on row number. Row # 2 has ST_NUM NaN but ST_NAME is LEXINGTON. Since the ST_NUM of LEXINGTON is given as 197 in row 1, so I will fill that NaN value with 197. In case of NaN value of ST_NUM of row 6, we find that ST_NUM hole 208 of WASHINGTON ST.

In [8]:
# Row wise data filling
df.loc[2,'ST_NUM'] = 197
df.loc[6,'ST_NUM'] = 208
df

PID  ST_NUM     ST_NAME OWN_OCCUPIED BEDROOMS    BATH SQ_FT
0  100001000.0   104.0      PUTNAM            Y        3       1  1000
1  100002000.0   197.0   LEXINGTON            N        3     1.5    --
2  100003000.0   197.0   LEXINGTON            N      NaN       1   850
3  100004000.0   201.0    BERKELEY           12        1     NaN   700
4  100005000.0   203.0    BERKELEY            Y        3       2  1600
5  100006000.0   207.0    BERKELEY            Y      NaN       1   800
6  100007000.0   208.0  WASHINGTON          NaN        2  HURLEY   950
7  100008000.0   213.0     TREMONT            Y        1       1   NaN
8  100009000.0   215.0     TREMONT            Y       na       2  1800

### Unwanted Value Handeling

The OWN_OCCUPIED column contains an unwanted value named 12 in addition to NaN, which is outside the value domain of this column. The values in this column must be Y or N. In the following method we will convert the unwanted (in this case numeric) value of this column to NaN.

In [9]:
#unwanted value treatment
cnt=0
for row in df['OWN_OCCUPIED']:
    try:
        int(row)
        df.loc[cnt, 'OWN_OCCUPIED']=np.nan
    except ValueError:
        pass
    cnt+=1
df

PID  ST_NUM     ST_NAME OWN_OCCUPIED BEDROOMS    BATH SQ_FT
0  100001000.0   104.0      PUTNAM            Y        3       1  1000
1  100002000.0   197.0   LEXINGTON            N        3     1.5    --
2  100003000.0   197.0   LEXINGTON            N      NaN       1   850
3  100004000.0   201.0    BERKELEY          NaN        1     NaN   700
4  100005000.0   203.0    BERKELEY            Y        3       2  1600
5  100006000.0   207.0    BERKELEY            Y      NaN       1   800
6  100007000.0   208.0  WASHINGTON          NaN        2  HURLEY   950
7  100008000.0   213.0     TREMONT            Y        1       1   NaN
8  100009000.0   215.0     TREMONT            Y       na       2  1800

BEDROOMS, BATH and SQ_FT These three columns are supposed to have only numerical values but we can see that these columns have some unwanted non-numeric values. So we will first convert them to NaN.

In [10]:
df['BEDROOMS'] = pd.to_numeric(df['BEDROOMS'], errors='coerce')
df['BATH'] = pd.to_numeric(df['BATH'], errors='coerce')
df['SQ_FT'] = pd.to_numeric(df['SQ_FT'], errors='coerce')
df

PID  ST_NUM     ST_NAME OWN_OCCUPIED  BEDROOMS  BATH   SQ_FT
0  100001000.0   104.0      PUTNAM            Y       3.0   1.0  1000.0
1  100002000.0   197.0   LEXINGTON            N       3.0   1.5     NaN
2  100003000.0   197.0   LEXINGTON            N       NaN   1.0   850.0
3  100004000.0   201.0    BERKELEY          NaN       1.0   NaN   700.0
4  100005000.0   203.0    BERKELEY            Y       3.0   2.0  1600.0
5  100006000.0   207.0    BERKELEY            Y       NaN   1.0   800.0
6  100007000.0   208.0  WASHINGTON          NaN       2.0   NaN   950.0
7  100008000.0   213.0     TREMONT            Y       1.0   1.0     NaN
8  100009000.0   215.0     TREMONT            Y       NaN   2.0  1800.0

### Fill out NAN using Mode

In [11]:
df['OWN_OCCUPIED'].fillna(df['OWN_OCCUPIED'].mode()[0], inplace=True)
df

PID  ST_NUM     ST_NAME OWN_OCCUPIED  BEDROOMS  BATH   SQ_FT
0  100001000.0   104.0      PUTNAM            Y       3.0   1.0  1000.0
1  100002000.0   197.0   LEXINGTON            N       3.0   1.5     NaN
2  100003000.0   197.0   LEXINGTON            N       NaN   1.0   850.0
3  100004000.0   201.0    BERKELEY            Y       1.0   NaN   700.0
4  100005000.0   203.0    BERKELEY            Y       3.0   2.0  1600.0
5  100006000.0   207.0    BERKELEY            Y       NaN   1.0   800.0
6  100007000.0   208.0  WASHINGTON            Y       2.0   NaN   950.0
7  100008000.0   213.0     TREMONT            Y       1.0   1.0     NaN
8  100009000.0   215.0     TREMONT            Y       NaN   2.0  1800.0

In [12]:
#Group By parameter check
df.groupby('BEDROOMS')['SQ_FT'].median()

BEDROOMS
1.0     700.0
2.0     950.0
3.0    1300.0
Name: SQ_FT, dtype: float64

In [13]:
# Filling Null with group by vparameter
df['SQ_FT'] = df['SQ_FT'].fillna(df.groupby('BEDROOMS')['SQ_FT'].transform('median'))
df['SQ_FT'] = df['SQ_FT'].fillna(df['SQ_FT'].median())
df

PID  ST_NUM     ST_NAME OWN_OCCUPIED  BEDROOMS  BATH   SQ_FT
0  100001000.0   104.0      PUTNAM            Y       3.0   1.0  1000.0
1  100002000.0   197.0   LEXINGTON            N       3.0   1.5  1300.0
2  100003000.0   197.0   LEXINGTON            N       NaN   1.0   850.0
3  100004000.0   201.0    BERKELEY            Y       1.0   NaN   700.0
4  100005000.0   203.0    BERKELEY            Y       3.0   2.0  1600.0
5  100006000.0   207.0    BERKELEY            Y       NaN   1.0   800.0
6  100007000.0   208.0  WASHINGTON            Y       2.0   NaN   950.0
7  100008000.0   213.0     TREMONT            Y       1.0   1.0   700.0
8  100009000.0   215.0     TREMONT            Y       NaN   2.0  1800.0

BedROOM's NaN has been filled by 1 bedroom in a small apartment of 950 square feet and 3 bedrooms in a large apartment of 1000 square feet according to the size of the apartment.

In [14]:
df

PID  ST_NUM     ST_NAME OWN_OCCUPIED  BEDROOMS  BATH   SQ_FT
0  100001000.0   104.0      PUTNAM            Y       3.0   1.0  1000.0
1  100002000.0   197.0   LEXINGTON            N       3.0   1.5  1300.0
2  100003000.0   197.0   LEXINGTON            N       NaN   1.0   850.0
3  100004000.0   201.0    BERKELEY            Y       1.0   NaN   700.0
4  100005000.0   203.0    BERKELEY            Y       3.0   2.0  1600.0
5  100006000.0   207.0    BERKELEY            Y       NaN   1.0   800.0
6  100007000.0   208.0  WASHINGTON            Y       2.0   NaN   950.0
7  100008000.0   213.0     TREMONT            Y       1.0   1.0   700.0
8  100009000.0   215.0     TREMONT            Y       NaN   2.0  1800.0

In [15]:
df['BATH']=df['BATH'].fillna(method='bfill')
df

PID  ST_NUM     ST_NAME OWN_OCCUPIED  BEDROOMS  BATH   SQ_FT
0  100001000.0   104.0      PUTNAM            Y       3.0   1.0  1000.0
1  100002000.0   197.0   LEXINGTON            N       3.0   1.5  1300.0
2  100003000.0   197.0   LEXINGTON            N       NaN   1.0   850.0
3  100004000.0   201.0    BERKELEY            Y       1.0   2.0   700.0
4  100005000.0   203.0    BERKELEY            Y       3.0   2.0  1600.0
5  100006000.0   207.0    BERKELEY            Y       NaN   1.0   800.0
6  100007000.0   208.0  WASHINGTON            Y       2.0   1.0   950.0
7  100008000.0   213.0     TREMONT            Y       1.0   1.0   700.0
8  100009000.0   215.0     TREMONT            Y       NaN   2.0  1800.0

#### Assignments

1. Find out the mising values from the entire columns
2. Find out the missing values from the last three columns, e.g., Bedrooms, bath and SQ_FT
3. Remove the noise from these columns
4. Use the above methods, e.g., drop the entire columns containing the null values and see whats going on or fill out through median/mode/previous digit/any other advance preprocessing method.
5. Check out the unwanted values from the columns
6. Make your observation
7. Experiment through different dataset, do collect from Kaggle/any other repositories.